src/meta/expr.lean

```diff
@@ -580,6 +580,37 @@ meta def to_implicit_binder : expr → expr
 | (pi n _ d b) := pi n binder_info.implicit d b
 | e  := e
 
+/--  Takes two `expr`s `f` and `x`, where `f` represents a binary operation.  Decomposes the
+second expression `x` into repeated applications of `f`, returning a list of the atoms of the
+matching process.
+
+See the related `expr.list_summands` and `expr.list_factors` for versions specialized to
+addition and multiplication. -/
+meta def list_binary_operands (f : expr) : expr → list expr
+| x@(expr.app (expr.app g a) b) :=
+  if f =ₐ g then a.list_binary_operands ++ b.list_binary_operands else [x]
```

Comment:
Does this correctly match `@has_add.add add_zero_class.to_has_add` with `@has_add.add add_semigroup.to_has_add` like `unify` did in my version?

Reply:
No, see my comment at https://github.com/leanprover-community/mathlib/pull/14617#issuecomment-1155512283 . It won't match different instances. My hypothesis is it won't be an issue for this application, but we should keep the `unify` version around for cases where it is, and I could be proven wrong.

Reply:
I think that sadly this type of thing comes all all the time; it's very common to end up with difference (but defeq) instances in an expression.
If I understand correctly, rewriting by `mul_add` is going to introduce a `has_add` that goes via `distrib`.

Reply:
Here's a concrete example:
```lean
example (n) (h : n = 10) : 1 + 2 + 3 + 4 = n :=
begin
  rw ←add_zero 4,
  rw add_assoc 1,
  -- [1 + (2 + 3), 4 + 0]
  do {
    `(%%l = %%r) ← tactic.target,
    let e := list_summands l,
    tactic.trace e },
  exact h.symm
end
```
Here the `list_summands` is failing to actually split the additions apart as a user would expect